### ISSUE TITLE
[API - feature request] Be able to create text element inside a given rectangle programatically

### ISSUE THREAD

**Issue by edgar-g-q** [excalidraw/excalidraw#6514]
I use Excalidraw to programmatically create rectangles with text inside them. To achieve that, I use the function [updateScene](https://docs.excalidraw.com/docs/@excalidraw/excalidraw/api/props/ref#updatescene) and I pass, as scene data, two elements for each pair rectangle-text. One element is the rectangle, whose position (x, y) and dimensions (width and height) I decide. The second element is the text. If I want to put the text centered in the rectangle, I am forced to compute multiple text parameters on my own (x, y, width, height, baseline, lineHeight, among others).

My question: is there any way to automatically compute the text properties in order to have the text properly centered in the rectangle?

**Comment by dwelle** [comment]
I don't believe there's a public API yet.

@ad1992 as per recent discussions on this, I wanted to add that just exposing a utility to programmatically bind text to containers or redraw dimensions won't help here.

Doing `updateScene(data)` and then having to follow it up with a `bindTextToContainer()` would be at best not very nice, require at least an access to the id (so you'd need to pregenerate it), and at worst impossible if we ever make `updateScene` async.

It will also not work for cases of initializing the scene data.

---

So for `updateScene()` or init data we should detect & redraw containers (and related elements) ourselves.

**Comment by edgar-g-q** [comment]
Thanks for your quick response.

I would like to clarify that, in my case, when I call updateScene, I am already providing all the values (including the id, which I generate randomly) for both rectangles and text (basically, I analysed the data structure of the elements and I just create the new objects on my own). I can also do the binding by using the property "boundElements" on the rectangle. However, when doing this it looks like the positional and dimensional parameters of the text are not recomputed after updating the scene.

It would be so convenient to have a function that can compute that for me, (I guess such a function would need to take into account the lenght of the text, the font family, font size, and other parameters). Maybe I could import that function from excalidraw package.

**Comment by ad1992** [comment]
> I don't believe there's a public API yet.
> 
> @ad1992 as per recent discussions on this, I wanted to add that just exposing a utility to programmatically bind text to containers or redraw dimensions won't help here.
> 
> Doing `updateScene(data)` and then having to follow it up with a `bindTextToContainer()` would be at best not very nice, require at least an access to the id (so you'd need to pregenerate it), and at worst impossible if we ever make `updateScene` async.
> 
> It will also not work for cases of initializing the scene data.
> 
> So for `updateScene()` or init data we should detect & redraw containers (and related elements) ourselves.

@dwelle my idea is that we expose `bindTextToContainer` which takes a container element and a text(string). We bind it and we return both the elements, now host is responsible for calling `updateScene` with those elements so its reflected in the scene.

**Comment by dwelle** [comment]
@ad1992 yeah, but like I said above, that's not good DX for when either seeding a component (`initialData`) or creating elements programmatically (`updateScene`). We need a better workflow than that.

**Comment by ad1992** [comment]
> @ad1992 yeah, but like I said above, that's not good DX for when either seeding a component (`initialData`) or creating elements programmatically (`updateScene`). We need a better workflow than that.

@dwelle I didn't get it, when user will be passing the elements to `updateScene`, it would be drawn automatically by Excal which is what we want. These elements will be pure utilities responsible for generating the elements and returning it and nothing more than that, once element generated, it's responsibility of host to call the api which is responsible for updating the drawings in the scene.

**Comment by dwelle** [comment]
The goal is to _at the very least_ make `updateScene()`, when being passed this:

```js
{
  elements: [
    { type: "rectangle", id: "rectangle1", boundElements: [{type: "text", id: "text1"}] },
    { type: "text", id: "text1", containerId: "rectangle1" }
  ]
}
```

to redraw/reposition the bound text correctly, which it currently does not.

But, the above alone isn't enough for a good DX when creating elements programmatically, because the users would need to 1) be aware of the binding implementation details (`boundElements` and `containerId`), and supply ids themselves.

For that, we'd want a higher-level DSL, such as:

```js
{
  elements: [
    { type: "rectangle", children: [{ type: "text" }] }
  ]
}
```

**Comment by ad1992** [comment]
closed by #6546
